src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +44,227 @@ class FunctionToStartOfInterval : public IFunction
     {
         const auto & time_column = arguments[0];
         const auto & interval_column = arguments[1];
-        const auto & time_zone = extractTimeZoneFromFunctionArguments(arguments, 2, 0);
-        auto result_column = dispatchForTimeColumn(time_column, interval_column, result_type, time_zone);
+
+        ColumnWithTypeAndName origin_column;
+        const bool has_origin_arg = (arguments.size() == 3 && isDateOrDate32OrDateTimeOrDateTime64(arguments[2].type)) || arguments.size() == 4;
+        if (has_origin_arg)
+            origin_column = arguments[2];
+
+        const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && isString(arguments[2].type))) ? 2 : 3;
+        const auto & time_zone = extractTimeZoneFromFunctionArguments(arguments, time_zone_arg_num, 0);
+
+        auto result_column = dispatchForTimeColumn(time_column, interval_column, origin_column, result_type, time_zone);
         return result_column;
     }
 
 private:
     ColumnPtr dispatchForTimeColumn(
-        const ColumnWithTypeAndName & time_column, const ColumnWithTypeAndName & interval_column, const DataTypePtr & result_type, const DateLUTImpl & time_zone) const
+        const ColumnWithTypeAndName & time_column, const ColumnWithTypeAndName & interval_column, const ColumnWithTypeAndName & origin_column, const DataTypePtr & result_type, const DateLUTImpl & time_zone) const
     {
-        const auto & from_datatype = *time_column.type.get();
+        const auto & time_column_type = *time_column.type.get();
+        const auto & time_column_col = *time_column.column.get();
 
-        if (isDateTime64(from_datatype))
+        if (isDateTime64(time_column_type))
         {
-            const auto * time_column_vec = checkAndGetColumn<ColumnDateTime64>(time_column.column.get());
-            auto scale = assert_cast<const DataTypeDateTime64 &>(from_datatype).getScale();
+            if (!isDateTime64(origin_column.type.get()) && origin_column.column != nullptr)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Datetime argument and origin argument for function {} must have the same type", getName());
+
+            const auto * time_column_vec = checkAndGetColumn<ColumnDateTime64>(time_column_col);
+            auto scale = assert_cast<const DataTypeDateTime64 &>(time_column_type).getScale();
 
             if (time_column_vec)
-                return dispatchForIntervalColumn(assert_cast<const DataTypeDateTime64 &>(from_datatype), *time_column_vec, interval_column, result_type, time_zone, scale);
+                return dispatchForIntervalColumn(assert_cast<const DataTypeDateTime64 &>(time_column_type), *time_column_vec, interval_column, origin_column, result_type, time_zone, scale);
         }
-        if (isDateTime(from_datatype))
+        else if (isDateTime(time_column_type))
```

Comment:
The previous comment about concatenating the conditions of course also applied to l. 234-235, l. 244-245 and l. 254-255.